### DISCUSSION TITLE
Where to find generated error reports? (/layers.../svm_err...md)

### DISCUSSION THREAD

**Discussion by tabiStein** [paketo-buildpacks/feedback#232] [Java Team]
I initially posted this question to the [Graal discussion board](https://github.com/oracle/graal/discussions/8006), but they suggested I come here. 

I'm trying to build a native image using the Spring Maven plugin with paketobuildpacks tiny builder. The build-image goal fails, telling me that there is a generated error report I should inspect. However, I don't see it anywhere on my computer (it doesn't seem to be a valid absolute path or relative path from my project).

Where should I be able to find this error report? Is it located in some Graal temp directory outside of my project?

```
[INFO]     [creator]     Failed generating '_my-main-class-name_' after 16.7s.
[INFO]     [creator]     
[INFO]     [creator]     The build process encountered an unexpected error:
[INFO]     [creator]     
[INFO]     [creator]     > java.lang.NoClassDefFoundError: javax/mail/MessagingException
[INFO]     [creator]     
**[INFO]     [creator]     Please inspect the generated error report at:
[INFO]     [creator]     /layers/paketo-buildpacks_native-image/native-image/svm_err_b_20231211T215514.128_pid232.md**
[INFO]     [creator]     
[INFO]     [creator]     If you are unable to resolve this problem, please file an issue with the error report at:
[INFO]     [creator]     https://graalvm.org/support
[INFO]     [creator]     unable to invoke layer creator
[INFO]     [creator]     unable to contribute native-image layer
[INFO]     [creator]     error running build
[INFO]     [creator]     exit status 1
[INFO]     [creator]     ERROR: failed to build: exit status 1
```

In case it's relevant, here is the command I'm running: mvn -Pnative spring-boot:build-image 
I copied the native profile from spring-boot-starter-parent, as follows:

```
<profile>
      <id>native</id>
      <build>
          <plugins>
            <plugin>
              <groupId>org.apache.maven.plugins</groupId>
              <artifactId>maven-jar-plugin</artifactId>
              <configuration>
                <archive>
                  <manifestEntries>
                    <Spring-Boot-Native-Processed>true</Spring-Boot-Native-Processed>
                  </manifestEntries>
                </archive>
              </configuration>
            </plugin>
            <plugin>
              <groupId>org.springframework.boot</groupId>
              <artifactId>spring-boot-maven-plugin</artifactId>
              <version>${spring-boot.version}</version>
              <configuration>
                <image>
                  <builder>paketobuildpacks/builder:tiny</builder>
                  <env>
                    <BP_NATIVE_IMAGE>true</BP_NATIVE_IMAGE>
                  </env>
                </image>
              </configuration>
              <executions>
                <execution>
                  <id>process-aot</id>
                  <goals>
                    <goal>process-aot</goal>
                  </goals>
                </execution>
              </executions>
            </plugin>
            <plugin>
              <groupId>org.graalvm.buildtools</groupId>
              <artifactId>native-maven-plugin</artifactId>
              <version>0.9.22</version>
              <configuration>
                <classesDirectory>${project.build.outputDirectory}</classesDirectory>
                <metadataRepository>
                  <enabled>true</enabled>
                </metadataRepository>
                <requiredVersion>22.3</requiredVersion>
              </configuration>
              <executions>
                <execution>
                  <id>add-reachability-metadata</id>
                  <goals>
                    <goal>add-reachability-metadata</goal>
                  </goals>
                </execution>
              </executions>
            </plugin>
          </plugins>
      </build>
    </profile>
```

**Comment by dmikusa** [comment]
Right now, there's not a lot you can do about that. The native image tool is writing the error report to the disk *inside* the build container. When the build is done, success or fail, it gets destroyed.

It looks like it's writing a text file, so maybe there's a way to make it just write that to STDOUT. The other option would be to attach a volume and have it write the files to the volume's location. Do you know if it's possible to change the location to which it writes these files? Either to STDOUT or to another spot on disk?

If none of those options work, then we can open an enhancement request to look for this build message and capture those files. We could then write to STDOUT or copy to a volume.

**Reply by tabiStein** [reply]
Thanks for your quick reply! Now I can see the error report in the container for a split second before the container shuts down and deletes it. The layers/ directory is apparently already a volume mounted to /var/lib/docker/volumes/pack-layers-xqlcejrsxl/_data, but it's immediately deleted as soon as the paketo container shuts down. Any idea if that's due to a paketobuildpacks/builder:tiny configuration, or should I investigate the inner workings of the Spring Boot Maven Plugin?

Regarding your idea to switch the native-image command output to standard out, I don't see anything for this in their documentation; they do have a "-H:TempDirectory=FileSystemPath" option that I am able to set with Paketo's "BP_NATIVE_IMAGE_BUILD_ARGUMENTS" environment variable, which lets you set a different directory for the error output ([documentation](https://www.graalvm.org/latest/reference-manual/native-image/overview/HostedvsRuntimeOptions/#debug-options)). Maybe there's a creative way to make use of that?

**Reply by tabiStein** [reply]
Aha! I figured out a solution in the meantime. The plugin offers an option for mount binding; these files aren't deleted on container deletion like the volumes in the /var/lib/docker directory.

Here's the plugin configuration for future reference:

```
              <configuration>
                <image>
                  <builder>paketobuildpacks/builder:tiny</builder>
                  <env>
                    <BP_NATIVE_IMAGE>true</BP_NATIVE_IMAGE>
                  </env>
                  <bindings>
                    <binding>/Users/admin/projects/native-image-output:/layers/paketo-buildpacks_native-image:rw</binding>
                  </bindings>
                </image>
              </configuration>
```

Thanks again for your help!

**Reply by dmikusa** [reply]
Excellent! I would suggest one minor adjustment. Instead of binding directly to the `/layers/...` directory, change that to a directory that does not exist in the container like `/build-data` and then use the Native Image argument `-H:TempDirectory=/build-data` to have the files written there.

If using `/layers/...` works for you great, but the buildpack writes stuff to that folder as well and there are certain assumptions made about it. By binding to that location, you are changing that directory and it may violate those assumptions and cause weird issues. Again, if it works great, but you still might want to consider the change above. If you see any weirdness try the change above and see if that resolves it.

**Reply by tabiStein** [reply]
Great suggestion, thanks again!
